[IMP] website: review `s_rating` snippet design

chgo-odoo opened the pull request:
This commit aims to review the `s_rating` snippet design to make it more in line with the new snippets design.

With the new snippets introduced in Odoo 18 and all the new possibilities unlocked for the user, it made perfectly sense to review the design of the `s_rating` snippet to make it fit as much scenario as possible.

task-4306631

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/186130.png)](https://mergebot.odoo.com/odoo/odoo/pull/186130)

chgo-odoo commented:
@stefanorigano, following the discussion on https://github.com/odoo/odoo/pull/183965 👋 

I have : 

- Removed the `-o` style from default
- Set the title to hidden by default
- Added the `role=""` attributes
- Used the `warning` and `border-color` colors for the icons
- Change a bit the size available for the icons since `3x` is really big and corner case IMO, `lg` and `2x` seems more generic and smooth as a progression

lebl-odoo commented:
Thx for reaching me out, I had a look and it suits me. Thx @chgo-odoo to change all outline icons!

Last thing, shouldn't we use another label for the stars, I find "Quality" a real use case and it doesn't help users to use it.
What do you think about "Rating score" instead of "Quality"?

chgo-odoo commented:
Thanks @lebl-odoo, in the end I went for Customer rating

stefanorigano commented on addons/website/static/src/snippets/s_rating/options.js:
```diff
@@ -126,17 +126,18 @@ options.registry.Rating = options.Class.extend({
         const $inactiveIcons = this.$target.find('.s_rating_inactive_icons > i');
         $activeIcons.removeClass().addClass(faClassActiveIcons);
         $inactiveIcons.removeClass().addClass(faClassInactiveIcons);
+        this.$target[0].setAttribute("aria-label", `${this.nbActiveIcons} out of ${this.nbTotalIcons} stars`);
```
Nice, almost there 👍 
Unfortunately this implementation wouldn't take in account other languages, but it's easy to address ;) 
You should import the `translation` module from `core`...

```js
import { _t } from "@web/core/l10n/translation";
```

... and use it when you need to output the string

```js
_t("Hello %s", chgo_name);
```

You can see it in action in `/web/static/src/core/file_upload/file_upload_progress_record.js`, but in a lot of other modules as well ;)

chgo-odoo commented:
Thanks @stefanorigano 🙏 

It should be good now

stefanorigano commented on addons/website/static/src/snippets/s_rating/options.js:
```diff
@@ -126,17 +127,21 @@ options.registry.Rating = options.Class.extend({
         $activeIcons.removeClass().addClass(faClassActiveIcons);
         $inactiveIcons.removeClass().addClass(faClassInactiveIcons);
+        this.$target[0].setAttribute("aria-label", _t("%(amount)s out of %(total)s stars", {
+            amount: this.nbActiveIcons,
```
If there is no rating (no yellow stars) , this code will output `"NaN"`  ;)

Suggestion:
```suggestion
            amount: this.nbActiveIcons || 0, 
```

chgo-odoo replied:
Thanks @stefanorigano 😅

stefanorigano reviewed (approved):
There's one little detail (note in the code) , but can be addressed during tech review.
The rest LGTM,  thanks @chgo-odoo 👍

chgo-odoo commented:
Hello @qsm-odoo, this was tested by @lebl-odoo as we reworked this snippet a bit more than what we usually do.

Could you have a look at it please? 🙏

chgo-odoo commented:
Hello @qsm-odoo 👋 

I rebased the branch so it's up-to-date, if someone could have a look at it, so we can proceed with https://github.com/odoo/odoo/pull/183965

Thanks !

qsm-odoo commented on addons/website/static/src/snippets/s_rating/options.js:
```diff
@@ -124,17 +125,21 @@ options.registry.Rating = options.Class.extend({
-            'fa-circle': 'fa-circle-o',
-            'fa-square': 'fa-square-o',
-            'fa-heart': 'fa-heart-o'
+            'fa-star': 'fa-star',
+            'fa-thumbs-up': 'fa-thumbs-up',
+            'fa-circle': 'fa-circle',
+            'fa-square': 'fa-square',
+            'fa-heart': 'fa-heart'
```
I don't know the point of this change but in any case, this is not good: creating an object whose keys are equal to its values... why?

qsm-odoo commented on addons/website/views/snippets/s_rating.xml:
```diff
@@ -2,17 +2,17 @@
-                <i class="fa fa-star"/>
-                <i class="fa fa-star"/>
+            <span class="s_rating_active_icons" style="color: var(--warning)">
+                <i class="fa fa-star" role="presentation"/>
+                <i class="fa fa-star" role="presentation"/>
+                <i class="fa fa-star" role="presentation"/>
```
Is this presentation role kept when changing the score/number of icons?

qsm-odoo commented on addons/website/views/snippets/s_rating.xml:
```diff
@@ -2,17 +2,17 @@
-                <i class="fa fa-star"/>
-                <i class="fa fa-star"/>
-                <i class="fa fa-star"/>
+            <span class="s_rating_active_icons" style="color: var(--warning)">
```
Is warning a good idea? :thinking: Maybe hardcoding the color would be better no? Or at best use `var(--yellow)` (for once bootstrap "plain colors" might make sense?)

qsm-odoo commented on addons/website/views/snippets/s_rating.xml:
```diff
@@ -48,13 +48,13 @@
                 <we-button data-select-class="" title="Small" data-img="/website/static/src/img/snippets_options/size_small.svg"/>
-                <we-button data-select-class="fa-2x" title="Medium" data-img="/website/static/src/img/snippets_options/size_medium.svg"/>
-                <we-button data-select-class="fa-3x" title="Large" data-img="/website/static/src/img/snippets_options/size_large.svg"/>
+                <we-button data-select-class="fa-lg" title="Medium" data-img="/website/static/src/img/snippets_options/size_medium.svg"/>
+                <we-button data-select-class="fa-2x" title="Large" data-img="/website/static/src/img/snippets_options/size_large.svg"/>
```
That's ok, but it would not be too much to add 4th button for the 3x that we had.

And if you really don't want it, then you can then just *hide* it (not remove it) (example of this kind of stuff can be found with `shape_img_square_opt`). The point is that if old snippets where people used fa-3x are edited, they will be able to properly remove the fa-3x class (explain that with a comment next to the code hiding the button).

chgo-odoo commented:
Hello @qsm-odoo, thanks for your review, sorry it took me some time to understand what was wrong with the JS part but indeed, we could remove that whole piece of code...

I adapted the PR with your comments, it should be good now

qsm-odoo commented on addons/website/static/src/snippets/s_rating/options.js:
```diff
@@ -123,18 +131,12 @@ options.registry.Rating = options.Class.extend({
-        const $inactiveIcons = this.$target.find('.s_rating_inactive_icons > i');
-        $activeIcons.removeClass().addClass(faClassActiveIcons);
-        $inactiveIcons.removeClass().addClass(faClassInactiveIcons);
+        const faClassIcons = (this.iconType === "custom") ? this.faClassActiveCustomIcons : 'fa ' + this.iconType;
+        const $ratingIcons = this.$target.find('.s_rating_icons i');
+        $ratingIcons.removeClass().addClass(faClassIcons);
+        this.$target[0].setAttribute("aria-label", _t("%(amount)s out of %(total)s stars", {
+            amount: this.nbActiveIcons || 0,
+            total: this.nbTotalIcons,
+        }));
```
@chgo-odoo Ask for a review once you will have retested everything please... this is just broken:

- Set an active icon to some specific icon -> change the score of the rating -> all icons, including the non-active ones change to the "active" icon choice
- Set an inactive icon to some specific icon -> change the score of the rating -> you lose the icon choice
- ...

I don't think you have to change anything in this JS code besides removing the "active -> inactive icon map"...

qsm-odoo reviewed (changes requested):
Commit message: max 72 characters per line

chgo-odoo commented:
Apologies for the waste of time @qsm-odoo, should've test deeper to catch all the possible scenarios.

I tested all of the above, everything is working as expected and I also handled the `3x` option as you suggested.

qsm-odoo reviewed (approved):
Added this diff: https://github.com/odoo/odoo/compare/a835e087f7f55a7bbeb155db4406655d15c20ede..0d64efe15b1bb7fe7c8fb00978cc416f61429987, adding the comment I asked + fixed an UI issue.

Also fixed the commit message:
![image](https://github.com/user-attachments/assets/4f1541c4-84be-477d-a153-c1722a33823e)

@robodoo r+

robodoo commented:
@chgo-odoo @qsm-odoo staging failed: ci/runbot on 3742873f47393577144dd2ff5ae4775124683407 (view more at https://runbot.odoo.com/runbot/build/73130956)

qsm-odoo commented:
@robodoo retry

Feyensv commented:
@qsm-odoo probably the cause of failed stagings:
https://runbot.odoo.com/runbot/build/73160055

Could you take a look and maybe r- ? :pray:

tde-banana-odoo commented:
@robodoo r-

chgo-odoo commented:
Thanks for the heads-up @tde-banana-odoo @Feyensv, I missed that test adaptation 🤦

qsm-odoo commented:
Sorry, I saw that 
![image](https://github.com/user-attachments/assets/837a9a7f-b150-4786-bbf5-d6e0139d2153)
then just `retry` because it did not seem to make sense for this PR.

We should really move that test from the theme repo to the website app somehow so it is also executed with website only somehow... annoying.

chgo-odoo commented:
Added a link to the design-themes PR and Runbot is green.

qsm-odoo reviewed (approved):
@robodoo r+
